net/ip.c

```diff
@@ -0,0 +1,448 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+#include "lib/assert.h"
+#include "lib/errno.h"
+#include "net/net_internal.h"
+#include <arpa/inet.h>          /* inet_pton, htons */
+#include <netdb.h>              /* hostent */
+#include <stdlib.h>             /* atoi */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_NET
+#include "lib/trace.h"
+
+static const char *ip_family[M0_NET_IP_AF_MAX]      = { "inet",
+							"inet6",
+							"unix" };
+
+static const char *ip_protocol[M0_NET_IP_PROTO_MAX] = { "tcp",
+							"verbs",
+							"stream",
+							"dgram",
+							"o2ib",
+							"lo" };
+
+/**
+ * Bitmap of used transfer machine identifiers. 1 is for used and 0 is for free.
+ */
+static uint8_t ip_autotm[1024] = {};
+
+/**
+ * This function convert the ip to hostname/FQDN format.
+ * Here ip is string containing ip address with numbes-and-dot notation.
+ */
+static void m0_net_ip_to_hostname(char *ip, char *hostname)
+{
+	struct hostent *he;
+	struct in_addr  addr;
+	int             rc;
+
+	inet_aton(ip, &addr);
+	he = gethostbyaddr(&addr, sizeof(addr), AF_INET);
+	M0_LOG(M0_DEBUG, "ip: %s to hostname: %s\n",ip, he->h_name);
+	rc = snprintf(hostname, M0_NET_IP_STRLEN_MAX, "%s", he->h_name);
+
+	if (rc >= M0_NET_IP_STRLEN_MAX)
+		M0_LOG(M0_ERROR, "Hostname too long. Reqd: %d buf_len: %d", rc,
+		       M0_NET_IP_STRLEN_MAX);
+}
+
+/**
+ * This function decodes the inet format address.
+ *  The inet address format is of type
+ *    <family>:<type>:<ipaddr/hostname_FQDN>@<port>
+ *    for example: "inet:tcp:127.0.0.1@3000",
+ *                 "inet:stream:lanl.gov@23",
+ *                 "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663"
+ * Return value: 0 in case of success.
+ *               < 0 in case of error.
+ */
+static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	uint32_t     portnum;
+	int          shift = 0;
+	int          f;
+	int          s;
+	int          rc;
+	char         ip[M0_NET_IP_STRLEN_MAX] = {};
+	char         port[M0_NET_IP_PORTLEN_MAX] = {};
+	char        *at;
+	const char  *ep_name = name;
+
+	for (f = 0; f < ARRAY_SIZE(ip_family); ++f) {
+		if (ip_family[f] != NULL) {
+			shift = strlen(ip_family[f]);
+			if (strncmp(ep_name, ip_family[f], shift) == 0)
+				break;
+		}
+	}
+	if (f >= ARRAY_SIZE(ip_family) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	for (s = 0; s < ARRAY_SIZE(ip_protocol); ++s) {
+		if (ip_protocol[s] != NULL) {
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(ep_name, ip_protocol[s], shift) == 0)
+				break;
+		}
+	}
+	if (s >= ARRAY_SIZE(ip_protocol) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
```

Comment:
A repeating pattern could be extracted into a separate function. It will reduce the amount of copy-pasted code.
Example:

```c
int prefix_consume(const char **pstr, const char *prefixes[], int nr_prefixes, char delim, int *index);
/* ... */
{
/* ... */
rc = prefix_consume(&ep_name, ip_family, ARRAY_SIZE(ip_family), ':', &f) ?:
prefix_consume(&ep_name, ip_protocol, ARRAY_SIZE(ip_protocol), ':', &s);
}
```

Reply:
new parse_prefix() added to avoid code duplication.